File: java/L100/LeetCode0135.java
package L100;

import java.util.Arrays;

public class LeetCode0135 {
    public int candy(int[] ratings) {
        final int N = ratings.length;
        int[] alloc = new int[N];
        alloc[0] = 1;
        for (int i = 1;i < N;++i) {
            if (ratings[i] > ratings[i - 1]) {
                alloc[i] = alloc[i - 1] + 1;
            } else {
                alloc[i] = 1;
            }
        }
        for (int i = N - 2;i >= 0;--i) {
            if (ratings[i] > ratings[i + 1]) {
                alloc[i] = Math.max(alloc[i], alloc[i + 1] + 1);
            } else {
                alloc[i] = Math.max(alloc[i], 1);
            }
        }
        return Arrays.stream(alloc).sum();
    }
}

File: cpp/L100/LeetCode0135.cc
#include <algorithm>
#include <vector>
using std::vector;

class Solution {
public:
    int candy(vector<int>& ratings) {
        int N = ratings.size();
        vector<int> count(N, 0);
        count[0] = 1;
        for (int i = 1;i < N;++i) {
            if (ratings[i] > ratings[i - 1]) {
                count[i] = count[i - 1] + 1;
            } else {
                count[i] = 1;
            }
        }
        int last{1};
        int ans{std::max(last, count[N - 1])};
        for (int i = N - 2;i >= 0;--i) {
            if (ratings[i] > ratings[i + 1]) {
                last++;
            } else {
                last = 1;
            }
            ans += std::max(count[i], last);
        }
        return ans;
    }
};